### ISSUE TITLE
Docker: No SIRF-Exercises

### ISSUE THREAD

**Issue by ashgillman** [SyneRBI/SIRF-SuperBuild#584]
The Docker README says that after following the TL;DR instructions that SIRF-Exercises will be available under ./devel, however they are not.

When running interactively to investigate:
```
> docker start -ai sirf
/opt/SIRF-SuperBuild ~
/devel /opt/SIRF-SuperBuild ~
cp: cannot create directory './SIRF-Exercises': Permission denied
/usr/bin/unzip
/usr/local/bin/service.sh: line 49: ./SIRF-Exercises/scripts/download_*.sh: No such file or directory
/opt/SIRF-SuperBuild ~
~
[I 12:16:02.109 NotebookApp] JupyterLab beta preview extension loaded from /opt/pyvenv/lib/python2.7/site-packages/jupyterlab
[I 12:16:02.109 NotebookApp] JupyterLab application directory is /opt/pyvenv/share/jupyter/lab
[I 12:16:02.113 NotebookApp] Serving notebooks from local directory: /devel

```

**Comment by KrisThielemans** [comment]
which docker image are you trying? We haven't been able to push to docker hub for a while now. Things should be better when you build it yourself.

the `sirf:latest` doesn't have `SIRF-Exercises`, but the `sirf:service` does.

I note a python 2.7 above. Hopefully that isn't a problem anymore.

**Comment by KrisThielemans** [comment]
I'm assuming #591 intends to fix this issue (it doesn't say so).

I wonder if you haven't misread the documentation. The TL;DR  tells you how to get a "service" image up, which includes the `SIRF-Exercises`. It clearly states to use `sirf-compose-server`.

#591 modifies the section for the CLI docker, which does *not* include the jupyter server, but also not the `SIRF-Exercises`. I guess there would be no harm in installing the `SIRF-Exercises` there already (but there seems to be no obvious way to run them in that CLI container).

Note that
```
docker start -ai sirf
```
in the "service" image will just get you in a session where the jupyter server is running. You can probably only abort it (or possibly move it to the background).

It might be possible/advisable to run the `jupyter notebook`  as a background process, such that people can do `docker start -ai sirf` and get a normal prompt as well, maybe reducing the confusion. Create an issue (or even try it) if you think that'd be a good idea.

In summary, I think this is a case of not reading the documentation properly, or badly written documentation. Feel free to improve (but not as in #591).

**Comment by ashgillman** [comment]
Regarding my initial issue - I did indeed build myself. But the issue was (I think) that I was trying to update the docker image using `docker-compose` not `sirf-compose`, per the instructions fixed in #591, and I don't think it was updating the image that `sirf-compose up` was creating the container from.

Python2:
I think this is related, I was accidentally using an old image for the container by not updating the correct one?

> I wonder if you haven't misread the documentation. The TL;DR tells you how to get a "service" image up, which includes the SIRF-Exercises. It clearly states to use sirf-compose-server.

I did indeed spend some time reading all the documentation.

I think the TL;DR is fine, it will get a new user up and running quickly. As I said above, ignore my initial report on this issue. I only had this issue because I had an old image already, but the TL;DR is for new users. Maybe we could add an instruction to first delete any SyneRBI images/containers?

The line I modified in #591 is the only place in the instructions that instruct how to re-build the image once you already have one (it seems `up` will build the first time, but not update your image if you have one, right?) There are no instructions on how to update the service image, but once the core image is updated the next time you run the service `up` it will see and update that one for you, so this is fine. But it didn't seem to be working as written, after #591 I seem to have the correct experience.


> Note that
> 
> ```
> docker start -ai sirf
> ```
> 
> in the "service" image will just get you in a session where the jupyter server is running. You can probably only abort it (or possibly move it to the background).

Yes, but its quite useful to run this to get the output for debugging. Maybe I didn't need to add the `-i`.

> It might be possible/advisable to run the jupyter notebook as a background process, such that people can do docker start -ai sirf and get a normal prompt as well, maybe reducing the confusion. Create an issue (or even try it) if you think that'd be a good idea.

Its good how it is now. People who want a prompt can use `docker exec -w /devel -ti sirf /bin/bash`, which is also outlined in the documentation.

**Comment by KrisThielemans** [comment]
> I did indeed spend some time reading all the documentation.

you must be one of the few :-) Thanks for that!
 
> I think the TL;DR is fine, it will get a new user up and running quickly. As I said above, ignore my initial report on this issue. I only had this issue because I had an old image already, but the TL;DR is for new users. Maybe we could add an instruction to first delete any SyneRBI images/containers?
>

I believe you are correct. `docker-compose up` doesn't seem to update. I haven't figured this one out yet so cannot write proper documentation... (I know how to force build and prevent build)
 
> There are no instructions on how to update the service image, 

you are correct. That's missing. However, you modified a line the "CLI" version (i.e. building `latest`, not `service`)

> but once the core image is updated the next time you run the service `up` it will see and update that one for you, so this is fine. 

here I'm rather lost. Did you mean the `core` image? Are you saying that if `core` or `sirf` is updated, then `sirf-compose-server up` will update it?

**Comment by ashgillman** [comment]
Okay, I'm starting to understand I think.

My proposed line, then should go in this section?
https://github.com/SyneRBI/SIRF-SuperBuild/blob/master/docker/README.md#creating-a-container-providing-a-linux-based-jupyter-server-with-sirf

I think it might help to have a blurb at the start of the CLI and Service sections to explain this, I'll draft something later today.

**Comment by KrisThielemans** [comment]
yup.

tricky to avoid duplication (there was far too much at some point) but still be clear
